google_guest_agent/snapshot_listener.go

```diff
@@ -0,0 +1,252 @@
+//  Copyright 2020 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"strconv"
+	"syscall"
+	"time"
+
+	pb "github.com/GoogleCloudPlatform/guest-agent/google_guest_agent/protos"
+	"github.com/GoogleCloudPlatform/guest-logging-go/logger"
+	"github.com/hashicorp/golang-lru"
+	"google.golang.org/grpc"
+)
+
+const testPreSnapshotScript = "preSnapshotScript.sh"
+const testPostSnapshotScript = "postSnapshotScript.sh"
+const testIP = "169.254.0.1"
+const prodAddress = "169.254.169.254:8081"
+
+var (
+	seenPreSnapshotoperationIds, _  = lru.New(128)
+	seenPostSnapshotoperationIds, _ = lru.New(128)
+)
+
+type snapshotConfig struct {
+	Timeout               int // seconds
+	ContinueOnScriptError bool
+	PreSnapshotScriptURL  string
+	PostSnapshotScriptURL string
+	Enabled               bool
+}
+
+type scriptsTimeoutError struct {
+	path string
+}
+
+func (e *scriptsTimeoutError) Error() string {
+	return fmt.Sprintf("script timed out at: %s", e.path)
+}
+
+type invalidSnapshotConfig struct {
+	msg string
+}
+
+func (e *invalidSnapshotConfig) Error() string {
+	return fmt.Sprintf("invalid config: %s", e.msg)
+}
+
+func getSnapshotConfig() (snapshotConfig, error) {
+	setConfig()
+	var conf snapshotConfig
+	conf.Timeout = config.Section("Snapshots").Key("timeout_in_seconds").MustInt(60)
+	conf.ContinueOnScriptError = config.Section("Snapshots").Key("continue_on_script_error").MustBool(false)
+	conf.PreSnapshotScriptURL = config.Section("Snapshots").Key("pre_snapshot_script").MustString("")
+	conf.PostSnapshotScriptURL = config.Section("Snapshots").Key("post_snapshot_script").MustString("")
+
+	if conf.PreSnapshotScriptURL == "" && conf.PostSnapshotScriptURL == "" {
+		msg := "neither pre or post snapshot script has been configured"
+		logger.Errorf("%s", msg)
+		return conf, &invalidSnapshotConfig{msg}
+	}
```

Comment:
just to be explicit with above feedback, this should be either logged or returned with the message defined inline